### PR TITLE
Fix #138 - s-replace-all

### LINKED ISSUE #138: s-replace-all case-issue
If the string contains a substring that is equal to one of the keys of the `replacements` alist, but with a different case, then  `Wrong type argument: stringp, nil` is output.

This is an example from #62 
```
(s-replace-all '(("BLAH" . "wah")
         ("BLOO" ."woo"))
           "BLAH BLOO BLEE")
```

which outputs 
```
"wah woo BLEE"
```

However, if the key "BLAH" is replaced by "Blah", I get the error from above

```
(s-replace-all '(("Blah" . "wah")
         ("BLOO" ."woo"))
           "BLAH BLOO BLEE")
```

### PR DESCRIPTION
**Pull request by jasonm23**
In #138 errors were thrown by the replacer
function finding a nil replacement.

We solve it by returning found the original text instead of a
replacement, when the replacement value is nil.

### CODE REVIEW

**Comment by jasonm23** [comment]
Hi @magnars can you take a quick look?

**Comment by magnars** [comment]
Is the problem here that the `"FOO"` regexp actually matches `"foo"`? Otherwise, why are we getting a match that isn't a match?

File: s.el [hunk 1]
```diff
@@ -405,9 +405,14 @@ This is a simple wrapper around the built-in `string-match-p'."
...
   "REPLACEMENTS is a list of cons-cells. Each `car` is replaced with `cdr` in S."
   (declare (pure t) (side-effect-free t))
-  (replace-regexp-in-string (regexp-opt (mapcar 'car replacements))
-                            (lambda (it) (s--aget replacements it))
-                            s t t))
+  (replace-regexp-in-string
+   (regexp-opt (mapcar 'car replacements))
+   (lambda (it)
+     (let ((replacement (s--aget replacements it)))
+       (if (null replacement)
```
Comment (jasonm23) [inline s.el:412]: #152 as you can see we're just doing null safe.

**Comment by jasonm23** [comment]
Hi @magnars as it is, the lambda will return a `nil` when there's no correct case match, but replace will still call it.  So we need to handle that in the lambda.

**Comment by magnars** [comment]
Yes, due to `case-fold-search`. Shouldn't we instead do `(let ((case-fold-search nil)) ...)` around the search?

**Comment by jasonm23** [comment]
Then if someone wants case fold?

**Comment by magnars** [comment]
The signature of `s-replace-all` would then require that they enumerate all the cases, and everything would be working as intended.

**Comment by jasonm23** [comment]
If they don't they will get an this unexpected error.

Not sure what you're advocating here.  Is the fix a bad idea?

**Comment by magnars** [comment]
There is probably some misunderstanding here. I'm just advocating a better fix. :) Setting the dynamically bound `case-fold-search` temporarily to `nil` while we do the search, will remove this error, since the search will then only match whatever is in the list of replacements. It will no longer match any case - only the ones that are actually in the replacement table.

**Comment by jasonm23** [comment]
Ok, however if it's called with case-fold explicitly set to `t` (say in an enclosing `let`) wouldn't that be a problem.

**Comment by magnars** [comment]
No, the function signature only allows exact matches anyway. From one string to another.

**Comment by jasonm23** [comment]
Ok sounds good. :+1: just want to be sure

**Pull request merged**